File: dotnet/tests/HydraPayments.Sdk.Tests/ModelTest.cs
using System.Text.Json;
using System.Text.Json.Serialization;
using HydraPayments.Sdk.Models;
using Xunit;

namespace HydraPayments.Sdk.Tests;

public class ModelTest
{
    private static readonly JsonSerializerOptions Options = new()
    {
        PropertyNamingPolicy = JsonNamingPolicy.SnakeCaseLower,
        DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull
    };

    [Fact]
    public void Account_Serialization()
    {
        var account = new Account
        {
            Id = "acc_123",
            OwnerId = "owner_456",
            AccountType = "business",
            Currency = "GBP",
            Balance = "1000.00",
            CreatedAt = "2024-01-01T00:00:00Z"
        };

        var json = JsonSerializer.Serialize(account, Options);
        Assert.Contains("\"owner_id\":\"owner_456\"", json);
        Assert.Contains("\"account_type\":\"business\"", json);
        Assert.DoesNotContain("\"metadata\"", json);
        Assert.DoesNotContain("\"updated_at\"", json);
    }

    [Fact]
    public void Account_Deserialization()
    {
        var json = "{\"id\":\"acc_123\",\"owner_id\":\"owner_456\",\"account_type\":\"business\",\"currency\":\"GBP\",\"balance\":\"500.00\",\"metadata\":\"extra\",\"created_at\":\"2024-01-01T00:00:00Z\",\"updated_at\":\"2024-01-02T00:00:00Z\"}";
        var account = JsonSerializer.Deserialize<Account>(json, Options)!;

        Assert.Equal("acc_123", account.Id);
        Assert.Equal("owner_456", account.OwnerId);
        Assert.Equal("business", account.AccountType);
        Assert.Equal("GBP", account.Currency);
        Assert.Equal("500.00", account.Balance);
        Assert.Equal("extra", account.Metadata);
        Assert.Equal("2024-01-01T00:00:00Z", account.CreatedAt);
        Assert.Equal("2024-01-02T00:00:00Z", account.UpdatedAt);
    }

    [Fact]
    public void Transaction_Deserialization()
    {
        var json = "{\"id\":\"txn_123\",\"source_account_id\":\"acc_1\",\"dest_account_id\":\"acc_2\",\"amount\":\"100.00\",\"currency\":\"GBP\",\"status\":\"completed\",\"transaction_type\":\"transfer\",\"reference\":\"ref_1\",\"description\":\"payment\",\"previous_state_hash\":\"abc123\",\"created_at\":\"2024-01-01T00:00:00Z\",\"updated_at\":\"2024-01-02T00:00:00Z\"}";
        var tx = JsonSerializer.Deserialize<Transaction>(json, Options)!;

        Assert.Equal("txn_123", tx.Id);
        Assert.Equal("acc_1", tx.SourceAccountId);
        Assert.Equal("acc_2", tx.DestAccountId);
        Assert.Equal("100.00", tx.Amount);
        Assert.Equal("GBP", tx.Currency);
        Assert.Equal("completed", tx.Status);
        Assert.Equal("transfer", tx.TransactionType);
        Assert.Equal("ref_1", tx.Reference);
        Assert.Equal("payment", tx.Description);
        Assert.Equal("abc123", tx.PreviousStateHash);
        Assert.Equal("2024-01-01T00:00:00Z", tx.CreatedAt);
        Assert.Equal("2024-01-02T00:00:00Z", tx.UpdatedAt);
    }

    [Fact]
    public void Transaction_OptionalFieldsDefaultToNull()
    {
        var json = "{\"id\":\"txn_123\",\"amount\":\"50.00\",\"currency\":\"USD\",\"status\":\"pending\",\"transaction_type\":\"transfer\",\"created_at\":\"2024-01-01T00:00:00Z\"}";
        var tx = JsonSerializer.Deserialize<Transaction>(json, Options)!;

        Assert.Equal("txn_123", tx.Id);
        Assert.Null(tx.SourceAccountId);
        Assert.Null(tx.DestAccountId);
        Assert.Null(tx.Reference);
        Assert.Null(tx.Description);
        Assert.Null(tx.Metadata);
    }

    [Fact]
    public void Wallet_Deserialization()
    {
        var json = "{\"id\":\"wal_123\",\"owner_id\":\"owner_456\",\"wallet_type\":\"custodial\",\"chain\":\"ethereum\",\"address\":\"0xabc\",\"is_custodial\":true,\"encrypted_private_key\":\"enc_key\",\"created_at\":\"2024-01-01T00:00:00Z\",\"updated_at\":\"2024-01-02T00:00:00Z\"}";
        var wallet = JsonSerializer.Deserialize<Wallet>(json, Options)!;

        Assert.Equal("wal_123", wallet.Id);
        Assert.Equal("owner_456", wallet.OwnerId);
        Assert.Equal("custodial", wallet.WalletType);
        Assert.Equal("ethereum", wallet.Chain);
        Assert.Equal("0xabc", wallet.Address);
        Assert.True(wallet.IsCustodial);
        Assert.Equal("enc_key", wallet.EncryptedPrivateKey);
        Assert.Equal("2024-01-01T00:00:00Z", wallet.CreatedAt);
        Assert.Equal("2024-01-02T00:00:00Z", wallet.UpdatedAt);
    }

    [Fact]
    public void Wallet_NonCustodial()
    {
        var json = "{\"id\":\"wal_456\",\"owner_id\":\"owner_789\",\"wallet_type\":\"non-custodial\",\"chain\":\"solana\",\"address\":\"0xdef\",\"is_custodial\":false,\"created_at\":\"2024-01-01T00:00:00Z\"}";
        var wallet = JsonSerializer.Deserialize<Wallet>(json, Options)!;

        Assert.Equal("non-custodial", wallet.WalletType);
        Assert.False(wallet.IsCustodial);
        Assert.Null(wallet.EncryptedPrivateKey);
        Assert.Null(wallet.UpdatedAt);
    }

    [Fact]
    public void SplitEntry_Serialization()
    {
        var entry = new SplitEntry { AccountId = "acc_1", Percentage = 50.0 };
        var json = JsonSerializer.Serialize(entry, Options);

        Assert.Contains("\"account_id\":\"acc_1\"", json);
        Assert.Contains("\"percentage\":50", json);
    }

    [Fact]
    public void SplitRule_WithNestedSplits()
    {
        var json = "{\"id\":\"split_123\",\"transaction_id\":\"txn_123\",\"total\":\"100.00\",\"currency\":\"GBP\",\"splits\":[{\"account_id\":\"acc_1\",\"percentage\":60.0},{\"account_id\":\"acc_2\",\"percentage\":40.0}],\"sink_account_id\":\"acc_3\",\"status\":\"active\",\"created_at\":\"2024-01-01T00:00:00Z\"}";
        var split = JsonSerializer.Deserialize<SplitRule>(json, Options)!;

        Assert.Equal("split_123", split.Id);
        Assert.Equal("txn_123", split.TransactionId);
        Assert.Equal("100.00", split.Total);
        Assert.Equal("GBP", split.Currency);
        Assert.Equal(2, split.Splits.Count);
        Assert.Equal("acc_1", split.Splits[0].AccountId);
        Assert.Equal(60.0, split.Splits[0].Percentage);
        Assert.Equal("acc_2", split.Splits[1].AccountId);
        Assert.Equal(40.0, split.Splits[1].Percentage);
        Assert.Equal("acc_3", split.SinkAccountId);
        Assert.Equal("active", split.Status);
    }

    [Fact]
    public void CardInput_Serialization()
    {
        var card = new CardInput { Number = "4111111111111111", ExpMonth = 12, ExpYear = 2026, Cvc = "123" };
        var json = JsonSerializer.Serialize(card, Options);

        Assert.Contains("\"number\":\"4111111111111111\"", json);
        Assert.Contains("\"exp_month\":12", json);
        Assert.Contains("\"exp_year\":2026", json);
        Assert.Contains("\"cvc\":\"123\"", json);
    }

    [Fact]
    public void CreateTokenResponse_Deserialization()
    {
        var json = "{\"id\":\"tok_123\",\"card\":{\"brand\":\"Visa\",\"last4\":\"1111\",\"exp_month\":12,\"exp_year\":2026},\"created_at\":\"2024-01-01T00:00:00Z\"}";
        var response = JsonSerializer.Deserialize<CreateTokenResponse>(json, Options)!;

        Assert.Equal("tok_123", response.Id);
        Assert.Equal("Visa", response.Card.Brand);
        Assert.Equal("1111", response.Card.Last4);
        Assert.Equal(12, response.Card.ExpMonth);
        Assert.Equal(2026, response.Card.ExpYear);
    }

    [Fact]
    public void CreateIntentResponse_Deserialization()
    {
        var json = "{\"id\":\"pi_123\",\"status\":\"requires_payment_method\",\"amount\":2000,\"currency\":\"GBP\",\"client_secret\":\"sec_123\"}";
        var response = JsonSerializer.Deserialize<CreateIntentResponse>(json, Options)!;

        Assert.Equal("pi_123", response.Id);
        Assert.Equal("requires_payment_method", response.Status);
        Assert.Equal(2000, response.Amount);
        Assert.Equal("GBP", response.Currency);
        Assert.Equal("sec_123", response.ClientSecret);
    }

    [Fact]
    public void CreateRefundResponse_Deserialization()
    {
        var json = "{\"id\":\"ref_123\",\"status\":\"succeeded\",\"amount\":1000,\"charge\":\"ch_123\"}";
        var response = JsonSerializer.Deserialize<CreateRefundResponse>(json, Options)!;

        Assert.Equal("ref_123", response.Id);
        Assert.Equal("succeeded", response.Status);
        Assert.Equal(1000, response.Amount);
        Assert.Equal("ch_123", response.Charge);
    }

    [Fact]
    public void HealthResponse_Deserialization()
    {
        var json = "{\"status\":\"healthy\",\"version\":\"1.0.0\",\"database\":\"connected\"}";
        var health = JsonSerializer.Deserialize<HealthResponse>(json, Options)!;

        Assert.Equal("healthy", health.Status);
        Assert.Equal("1.0.0", health.Version);
        Assert.Equal("connected", health.Database);
        Assert.True(health.IsHealthy());
    }

    [Fact]
    public void HealthResponse_Unhealthy()
    {
        var json = "{\"status\":\"unhealthy\",\"version\":\"1.0.0\",\"database\":\"disconnected\"}";
        var health = JsonSerializer.Deserialize<HealthResponse>(json, Options)!;

        Assert.Equal("unhealthy", health.Status);
        Assert.False(health.IsHealthy());
    }

    [Fact]
    public void ApiResponse_Deserialization()
    {
        var json = "{\"success\":true,\"data\":{\"id\":\"acc_123\",\"owner_id\":\"owner_1\",\"account_type\":\"business\",\"currency\":\"GBP\",\"balance\":\"100.00\",\"created_at\":\"2024-01-01T00:00:00Z\"},\"error\":null}";
        var response = JsonSerializer.Deserialize<ApiResponse<Account>>(json, Options)!;

        Assert.True(response.Success);
        Assert.NotNull(response.Data);
        Assert.Equal("acc_123", response.Data!.Id);
        Assert.Null(response.Error);
    }

    [Fact]
    public void ApiResponse_WithError()
    {
        var json = "{\"success\":false,\"data\":null,\"error\":\"Something went wrong\"}";
        var response = JsonSerializer.Deserialize<ApiResponse<Account>>(json, Options)!;

        Assert.False(response.Success);
        Assert.Null(response.Data);
        Assert.Equal("Something went wrong", response.Error);
    }

    [Fact]
    public void OptionalFields_OmittedWhenNull()
    {
        var obj = new { Name = "test", Optional = (string?)null };
        var json = JsonSerializer.Serialize(obj, Options);

        Assert.Contains("\"name\":\"test\"", json);
        Assert.DoesNotContain("optional", json);
    }
}

File: java/src/test/java/com/hydrapayments/sdk/ModelTest.java
package com.hydrapayments.sdk;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.hydrapayments.sdk.model.*;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.*;

class ModelTest {

    private ObjectMapper mapper;

    @BeforeEach
    void setUp() {
        mapper = new ObjectMapper();
        mapper.configure(com.fasterxml.jackson.databind.DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
    }

    @Test
    void testAccountDeserialization() throws Exception {
        String json = "{\"id\":\"acc-1\",\"owner_id\":\"owner-123\",\"account_type\":\"personal\",\"currency\":\"GBP\",\"balance\":\"100.00\",\"created_at\":\"2024-01-01T00:00:00Z\"}";
        Account a = mapper.readValue(json, Account.class);
        assertEquals("acc-1", a.getId());
        assertEquals("owner-123", a.getOwnerId());
        assertEquals("personal", a.getAccountType());
        assertEquals("GBP", a.getCurrency());
        assertEquals("100.00", a.getBalance());
        assertEquals("2024-01-01T00:00:00Z", a.getCreatedAt());
        assertNull(a.getUpdatedAt());
    }

    @Test
    void testAccountWithOptionalFields() throws Exception {
        String json = "{\"id\":\"acc-1\",\"owner_id\":\"owner-1\",\"account_type\":\"company\",\"currency\":\"USD\",\"balance\":\"50.00\",\"metadata\":\"info\",\"created_at\":\"t1\",\"updated_at\":\"t2\"}";
        Account a = mapper.readValue(json, Account.class);
        assertEquals("info", a.getMetadata());
        assertEquals("t2", a.getUpdatedAt());
    }

    @Test
    void testTransactionDeserialization() throws Exception {
        String json = "{\"id\":\"tx-1\",\"source_account_id\":\"src-1\",\"dest_account_id\":\"dst-1\",\"amount\":\"99.99\",\"currency\":\"GBP\",\"status\":\"completed\",\"transaction_type\":\"transfer\",\"reference\":\"REF-1\",\"created_at\":\"t1\"}";
        Transaction t = mapper.readValue(json, Transaction.class);
        assertEquals("tx-1", t.getId());
        assertEquals("src-1", t.getSourceAccountId());
        assertEquals("dst-1", t.getDestAccountId());
        assertEquals("99.99", t.getAmount());
        assertEquals("completed", t.getStatus());
        assertEquals("REF-1", t.getReference());
    }

    @Test
    void testTransactionDefaultValues() throws Exception {
        String json = "{\"id\":\"tx-1\"}";
        Transaction t = mapper.readValue(json, Transaction.class);
        assertEquals("tx-1", t.getId());
        assertNull(t.getAmount());
        assertNull(t.getStatus());
    }

    @Test
    void testWalletDeserialization() throws Exception {
        String json = "{\"id\":\"w-1\",\"owner_id\":\"owner-1\",\"wallet_type\":\"custodial\",\"chain\":\"ethereum\",\"address\":\"0xabc\",\"is_custodial\":true,\"created_at\":\"t1\"}";
        Wallet w = mapper.readValue(json, Wallet.class);
        assertEquals("w-1", w.getId());
        assertEquals("ethereum", w.getChain());
        assertTrue(w.isCustodial());
    }

    @Test
    void testWalletNonCustodial() throws Exception {
        String json = "{\"id\":\"w-1\",\"owner_id\":\"owner-1\",\"wallet_type\":\"non-custodial\",\"chain\":\"solana\",\"address\":\"0xdef\",\"is_custodial\":false,\"encrypted_private_key\":null,\"created_at\":\"t1\"}";
        Wallet w = mapper.readValue(json, Wallet.class);
        assertEquals("non-custodial", w.getWalletType());
        assertFalse(w.isCustodial());
    }

    @Test
    void testSplitEntry() throws Exception {
        String json = "{\"account_id\":\"acc-1\",\"percentage\":60.0}";
        SplitEntry s = mapper.readValue(json, SplitEntry.class);
        assertEquals("acc-1", s.getAccountId());
        assertEquals(60.0, s.getPercentage(), 0.001);
    }

    @Test
    void testHealthResponse() throws Exception {
        String json = "{\"status\":\"healthy\",\"version\":\"1.0.0\",\"database\":\"connected\"}";
        HealthResponse h = mapper.readValue(json, HealthResponse.class);
        assertEquals("healthy", h.getStatus());
        assertEquals("1.0.0", h.getVersion());
        assertEquals("connected", h.getDatabase());
        assertTrue(h.isHealthy());
    }

    @Test
    void testHealthResponseUnhealthy() throws Exception {
        String json = "{\"status\":\"down\",\"version\":\"1.0.0\",\"database\":\"disconnected\"}";
        HealthResponse h = mapper.readValue(json, HealthResponse.class);
        assertFalse(h.isHealthy());
    }

    @Test
    void testApiResponseSuccess() throws Exception {
        String json = "{\"success\":true,\"data\":{\"id\":\"acc-1\"},\"error\":null}";
        ApiResponse<Account> r = mapper.readValue(json,
            mapper.getTypeFactory().constructParametricType(ApiResponse.class, Account.class));
        assertTrue(r.isSuccess());
        assertNull(r.getError());
    }

    @Test
    void testCardInputSerialization() throws Exception {
        CardInput c = new CardInput("4242424242424242", 12, 2027, "123");
        String json = mapper.writeValueAsString(c);
        assertTrue(json.contains("\"number\":\"4242424242424242\""));
        assertTrue(json.contains("\"exp_month\":12"));
        assertTrue(json.contains("\"exp_year\":2027"));
    }

    @Test
    void testCreateTokenResponse() throws Exception {
        String json = "{\"id\":\"tok_1\",\"card\":{\"brand\":\"visa\",\"last4\":\"4242\",\"exp_month\":12,\"exp_year\":2027},\"created_at\":\"t1\"}";
        CreateTokenResponse r = mapper.readValue(json, CreateTokenResponse.class);
        assertEquals("tok_1", r.getId());
        assertEquals("visa", r.getCard().getBrand());
        assertEquals("4242", r.getCard().getLast4());
    }

    @Test
    void testCreateIntentResponse() throws Exception {
        String json = "{\"id\":\"pi_1\",\"status\":\"requires_action\",\"amount\":5000,\"currency\":\"GBP\",\"client_secret\":\"secret_xxx\"}";
        CreateIntentResponse r = mapper.readValue(json, CreateIntentResponse.class);
        assertEquals("pi_1", r.getId());
        assertEquals("requires_action", r.getStatus());
        assertEquals(5000, r.getAmount());
        assertEquals("secret_xxx", r.getClientSecret());
    }

    @Test
    void testCreateRefundResponse() throws Exception {
        String json = "{\"id\":\"re_1\",\"status\":\"succeeded\",\"amount\":2500,\"charge\":\"ch_1\"}";
        CreateRefundResponse r = mapper.readValue(json, CreateRefundResponse.class);
        assertEquals("re_1", r.getId());
        assertEquals("succeeded", r.getStatus());
        assertEquals(2500, r.getAmount());
    }

    @Test
    void testCommissionResponse() throws Exception {
        String json = "{\"total_commission\":1550}";
        CommissionResponse r = mapper.readValue(json, CommissionResponse.class);
        assertEquals(1550, r.getTotalCommission());
    }

    @Test
    void testWebhookResponse() throws Exception {
        String json = "{\"received\":true}";
        WebhookResponse r = mapper.readValue(json, WebhookResponse.class);
        assertTrue(r.isReceived());
    }

    @Test
    void testWebhookResponseNotReceived() throws Exception {
        String json = "{\"received\":false}";
        WebhookResponse r = mapper.readValue(json, WebhookResponse.class);
        assertFalse(r.isReceived());
    }
}
